File: packages/agent/src/events/types/subagent.ts
/**
 * @fileoverview Subagent Events
 *
 * Events for subagent lifecycle management.
 */

import type { SessionId } from './branded.js';
import type { BaseEvent } from './base.js';
import type { TokenUsage } from './token-usage.js';

// =============================================================================
// Subagent Events
// =============================================================================

/** Spawn type for subagent */
export type SubagentSpawnType = 'subsession' | 'tmux' | 'fork';

/**
 * Subagent spawned event - emitted when a sub-agent is created
 */
export interface SubagentSpawnedEvent extends BaseEvent {
  type: 'subagent.spawned';
  payload: {
    /** Session ID of the spawned sub-agent */
    subagentSessionId: SessionId;
    /** Type of spawn (in-process vs tmux) */
    spawnType: SubagentSpawnType;
    /** Task/prompt given to the sub-agent */
    task: string;
    /** Model used by the sub-agent */
    model: string;
    /** Tools enabled for the sub-agent */
    tools?: string[];
    /** Skills loaded for the sub-agent */
    skills?: string[];
    /** Working directory for the sub-agent */
    workingDirectory: string;
    /** Tmux session name (only for tmux spawn type) */
    tmuxSessionName?: string;
    /** Maximum turns allowed */
    maxTurns?: number;
  };
}

/**
 * Subagent status update event - periodic status updates during execution
 */
export interface SubagentStatusUpdateEvent extends BaseEvent {
  type: 'subagent.status_update';
  payload: {
    /** Session ID of the sub-agent */
    subagentSessionId: SessionId;
    /** Current status */
    status: 'running' | 'paused' | 'waiting_input';
    /** Current turn number */
    currentTurn: number;
    /** Brief description of current activity */
    activity?: string;
    /** Token usage so far */
    tokenUsage?: TokenUsage;
  };
}

/**
 * Subagent completed event - emitted when sub-agent finishes successfully
 */
export interface SubagentCompletedEvent extends BaseEvent {
  type: 'subagent.completed';
  payload: {
    /** Session ID of the sub-agent */
    subagentSessionId: SessionId;
    /** Summary of what the sub-agent accomplished */
    resultSummary: string;
    /** Total turns taken */
    totalTurns: number;
    /** Total token usage */
    totalTokenUsage: TokenUsage;
    /** Duration in milliseconds */
    duration: number;
    /** Files modified by the sub-agent */
    filesModified?: string[];
    /** Final output/response from the sub-agent */
    finalOutput?: string;
  };
}

/**
 * Subagent failed event - emitted when sub-agent fails
 */
export interface SubagentFailedEvent extends BaseEvent {
  type: 'subagent.failed';
  payload: {
    /** Session ID of the sub-agent */
    subagentSessionId: SessionId;
    /** Error message */
    error: string;
    /** Error code if available */
    code?: string;
    /** Whether the error is recoverable */
    recoverable: boolean;
    /** Partial result if any work was completed before failure */
    partialResult?: string;
    /** Turn number when failure occurred */
    failedAtTurn?: number;
    /** Duration until failure in milliseconds */
    duration?: number;
  };
}

File: packages/agent/src/domains/session/event_store/types/payloads/subagent.rs
//! Subagent event payloads.

use serde::{Deserialize, Serialize};

use super::TokenUsage;

/// Payload for `subagent.spawned` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubagentSpawnedPayload {
    /// Child session ID.
    pub subagent_session_id: String,
    /// Spawn type.
    pub spawn_type: String,
    /// Task description.
    pub task: String,
    /// Model ID.
    pub model: String,
    /// Tools available.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tools: Option<Vec<String>>,
    /// Skills available.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub skills: Option<Vec<String>>,
    /// Working directory.
    pub working_directory: String,
    /// Tmux session name for tmux spawn type.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tmux_session_name: Option<String>,
    /// Maximum turns allowed.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_turns: Option<i64>,
}

/// Payload for `subagent.status_update` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubagentStatusUpdatePayload {
    /// Child session ID.
    pub subagent_session_id: String,
    /// Current status.
    pub status: String,
    /// Current turn number.
    pub current_turn: i64,
    /// Activity description.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub activity: Option<String>,
    /// Token usage so far.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token_usage: Option<TokenUsage>,
}

/// Payload for `subagent.completed` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubagentCompletedPayload {
    /// Child session ID.
    pub subagent_session_id: String,
    /// Result summary.
    pub result_summary: String,
    /// Total turns taken.
    pub total_turns: i64,
    /// Total token usage.
    pub total_token_usage: TokenUsage,
    /// Duration in milliseconds.
    pub duration: i64,
    /// Files modified.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub files_modified: Option<Vec<String>>,
    /// Final output text.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub final_output: Option<String>,
}

/// Payload for `subagent.failed` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubagentFailedPayload {
    /// Child session ID.
    pub subagent_session_id: String,
    /// Error message.
    pub error: String,
    /// Error code.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub code: Option<String>,
    /// Whether recoverable.
    pub recoverable: bool,
    /// Partial result.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub partial_result: Option<String>,
    /// Turn at which failure occurred.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub failed_at_turn: Option<i64>,
    /// Duration in milliseconds.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration: Option<i64>,
}
